### ISSUE TITLE
[RLlib]  Significant performance issue with curriculum learning when using a high number of rollout workers

### ISSUE THREAD

**Issue by mwulfman** [ray-project/ray#41894]
### What happened + What you expected to happen

- **What happened :** When I started using the curriculum training option for my training load — defining a curriculum capable env from my original env, and a curriculum function, and then adding the corresponding options to my algorithm configuration — I noticed a significant slow-down in my training iterations. In addition, the slow-down became more significant the more rollout workers I was using: about a 4s slow-down with 50 rollout workers, and about a 40s slow down with 500 rollout workers. For comparison, before using the curriculum, my training iterations were around 15s each with 500 rollout workers, but with the curriculum on, they were 55s! Creating a curriculum for a training load should have a minor impact on the duration of the training iterations. I also noticed that the slow-down increases as the training batch size increases. I am using `num_cpus_per_worker=1`.

-  **What I found :** The issue comes from bad handling of the `results` dictionary in the `step` method in [algorithms/algorithm.py](https://github.com/ray-project/ray/tree/master/rllib/algorithms/algorithm.py) when used in the function `fn(env, env_context, task_fn)` defined line `870`. From my understanding, as it currently stands, because each worker calls this function, each worker makes a copy of `results`. 
```
            def fn(env, env_context, task_fn):
                new_task = task_fn(results, env, env_context)
                cur_task = env.get_task()
                if cur_task != new_task:
                    env.set_task(new_task)

            fn = functools.partial(fn, task_fn=self.config.env_task_fn)
            self.workers.foreach_env_with_context(fn)
```
-  **What has worked :** There are two things that I did that solved the issue for my use case :

1. Putting the results in Ray’s distributed shared-memory object store. We can now retrieve the `results` with `ray.get` in `fn` : 

```
            results_ref = ray.put(results)
            def fn(env, env_context, task_fn):
                new_task = task_fn(ray.get(results_ref), env, env_context)
                cur_task = env.get_task()
                if cur_task != new_task:
                    env.set_task(new_task)

            fn = functools.partial(fn, task_fn=self.config.env_task_fn)
            self.workers.foreach_env_with_context(fn)
            del results_ref
 ``` 

This has worked really well on my use case — with many rollout workers — and divided the slow-down by almost 4-fold. However, for use cases with few rollout workers and a large num_env/num_workers ratio this can potentially result in worse performance. From several trials I made with few rollout workers and a large num_env/num_workers ratio, the curriculum update became slower but was still very acceptable in absolute. Tests would need to be conducted depending on your use case.

2.  The curriculum function `task_fn` doesn't need the whole `results` dictionary for most use cases.  I would suggest having an option to filter the `results` before using it `fn`, in conjunction (or not) with the previous suggestions. Here's an example:

   
```
            curriculum_metrics = ("episode_len_mean", "episode_reward_mean", "custom_metrics")
            results_ref = ray.put({key: value for key, value in results.items() if key in curriculum_metrics})
            def fn(env, env_context, task_fn):
                new_task = task_fn(ray.get(results_ref), env, env_context)
                cur_task = env.get_task()
                if cur_task != new_task:
                    env.set_task(new_task)

            fn = functools.partial(fn, task_fn=self.config.env_task_fn)
            self.workers.foreach_env_with_context(fn)
            del results_ref
```

This had provided an additional division of the slow-down by a factor of 10 for my use case. This latter number varies mainly with the number of variables stored in the`results`dictionary and with the training batch size. There is probably a cleaner way to define the `curriculum_metrics` set. Adding a field for curriculum metrics in the `.environment` algorithm configuration could be a possibility.

On my use case, I was able to go from a 40s slow-down to a 0.5s slow-down (which corresponds more what I was expecting originally).

**Remarks**
 - To time the slow-downs, I've simply timed the time spent in the `if` branch line `862`to `877` as follows:

```
        if self.config.env_task_fn is not None:
            if not callable(self.config.env_task_fn):
                raise ValueError(
                    "`env_task_fn` must be None or a callable taking "
                    "[train_results, env, env_ctx] as args!"
                )
            start = time.time()
            curriculum_metrics = ("episode_len_mean", "episode_reward_mean", "custom_metrics")
            results_ref = ray.put({key: value for key, value in results.items() if key in curriculum_metrics})
            def fn(env, env_context, task_fn):
                new_task = task_fn(ray.get(results_ref), env, env_context)
                cur_task = env.get_task()
                if cur_task != new_task:
                    env.set_task(new_task)

            fn = functools.partial(fn, task_fn=self.config.env_task_fn)
            self.workers.foreach_env_with_context(fn)
            del results_ref
            logging.error(f"Curriculum update time {time.time()-start:.3f}")
```

- My suggestions fix the main issue with little code change, however, I am not convinced of the role of each rollout workers to update the curriculum task of their own environments. Should it not be the role of a workers' manager to decide on the next curriculum task? I'm curious to know what you think about it.

### Versions / Dependencies

Python: 3.10.10
ray: 2.8.0
OS: Ubuntu 22.04

Other requirements (notably to run reproduction script):
absl-py==2.0.0
aiosignal==1.3.1
attrs==23.1.0
autoflake==2.2.1
certifi==2023.11.17
charset-normalizer==3.3.2
click==8.1.7
cloudpickle==3.0.0
contourpy==1.2.0
cycler==0.12.1
decorator==5.1.1
distlib==0.3.7
dm-tree==0.1.8
Farama-Notifications==0.0.4
filelock==3.13.1
fonttools==4.45.1
frozenlist==1.4.0
fsspec==2023.10.0
gast==0.5.4
grpcio==1.59.3
gymnasium==0.29.1
idna==3.5
imageio==2.33.0
Jinja2==3.1.2
jsonschema==4.20.0
jsonschema-specifications==2023.11.1
kiwisolver==1.4.5
lazy_loader==0.3
lz4==4.3.2
markdown-it-py==3.0.0
MarkupSafe==2.1.3
matplotlib==3.8.2
mdurl==0.1.2
mpmath==1.3.0
msgpack==1.0.7
networkx==3.2.1
numpy==1.26.2
packaging==23.2
pandas==2.1.3
Pillow==10.1.0
platformdirs==4.0.0
protobuf==4.25.1
pyarrow==14.0.1
pyflakes==3.1.0
pygame==2.5.2
Pygments==2.17.2
pyparsing==3.1.1
python-dateutil==2.8.2
pytz==2023.3.post1
PyYAML==6.0.1
ray==2.8.0
referencing==0.31.0
requests==2.31.0
rich==13.7.0
rpds-py==0.13.1
scikit-image==0.22.0
scipy==1.11.4
six==1.16.0
sympy==1.12
tabulate==0.9.0
tensorboardX==2.6.2.2
tensorflow-probability==0.23.0
tifffile==2023.9.26
tomli==2.0.1
torch==2.1.1
typer==0.9.0
typing_extensions==4.8.0
tzdata==2023.3
urllib3==2.1.0
virtualenv==20.24.7

### Reproduction script

Add the timer lines described above in [algorithms/algorithm.py](https://github.com/ray-project/ray/tree/master/rllib/algorithms/algorithm.py) first.

```
from typing import Dict
import time
import random

import numpy as np
import gymnasium as gym
from gymnasium.spaces import Discrete, Box

import ray
from ray.tune.registry import register_env
from ray.rllib.utils.framework import try_import_torch
from ray.rllib.algorithms.ppo import PPOConfig
from ray.rllib.algorithms.callbacks import DefaultCallbacks
from ray.rllib.env.apis.task_settable_env import TaskSettableEnv, TaskType
from ray.rllib.env.env_context import EnvContext
from ray.rllib.env import BaseEnv
from ray.rllib.utils.annotations import override
from ray.rllib.evaluation import Episode, RolloutWorker
from ray.rllib.policy import Policy
from ray.rllib.policy.sample_batch import SampleBatch

torch, nn = try_import_torch()

class SimpleCorridor(gym.Env):
    """Example of a custom env in which you have to walk down a corridor.

    You can configure the length of the corridor via the env config."""

    def __init__(self, config: EnvContext):
        self.end_pos = config["corridor_length"]
        self.cur_pos = 0
        self.action_space = Discrete(2)
        self.observation_space = Box(0.0, self.end_pos, shape=(1,), dtype=np.float32)
        # Set the seed. This is only used for the final (reach goal) reward.
        self.reset(seed=config.worker_index * config.num_workers)

    def reset(self, *, seed=None, options=None):
        random.seed(seed)
        self.cur_pos = 0
        return [self.cur_pos], {}

    def step(self, action):
        assert action in [0, 1], action
        if action == 0 and self.cur_pos > 0:
            self.cur_pos -= 1
        elif action == 1:
            self.cur_pos += 1
        done = truncated = self.cur_pos >= self.end_pos

        info = {
            "current_position": float(self.cur_pos),
            "action": float(action),
            "is_truncated": float(truncated)
        }
        additional_info = {f"data_{i}": float(i) for i in range(15)}
        info.update(additional_info)

        return (
            [self.cur_pos],
            1.0 if done else -0.1,
            done,
            truncated,
            info,
        )

class MyCallbacks(DefaultCallbacks):
    def on_episode_start(
        self,
        *,
        worker: RolloutWorker,
        base_env: BaseEnv,
        policies: Dict[str, Policy],
        episode: Episode,
        env_index: int,
        **kwargs,
    ):
        episode.user_data["current_position"] = []
        episode.hist_data["current_position"] = []
        episode.user_data["action"] = []
        episode.user_data["is_truncated"] = []

    def on_episode_step(
        self,
        *,
        worker: RolloutWorker,
        base_env: BaseEnv,
        policies: Dict[str, Policy],
        episode: Episode,
        env_index: int,
        **kwargs,
    ):
        # Make sure this episode is ongoing.
        assert episode.length > 0, (
            "ERROR: `on_episode_step()` callback should not be called right "
            "after env reset!"
        )
        cur_pos = episode.last_info_for()["current_position"]
        action =  episode.last_info_for()["action"]
        is_truncated =  episode.last_info_for()["is_truncated"]

        episode.user_data["current_position"].append(cur_pos)
        episode.user_data["action"].append(action)
        episode.user_data["is_truncated"].append(is_truncated)

    def on_episode_end(
        self,
        *,
        worker: RolloutWorker,
        base_env: BaseEnv,
        policies: Dict[str, Policy],
        episode: Episode,
        env_index: int,
        **kwargs,
    ):
        current_position = np.mean(episode.user_data["current_position"])
        episode.custom_metrics["current_position"] = current_position
        episode.hist_data["current_position"] = episode.user_data["current_position"]


    def on_train_result(self, *, algorithm, result: dict, **kwargs):
        result["callback_ok"] = True

        current_position = result["custom_metrics"]["current_position"]
        var = np.var(current_position)
        mean = np.mean(current_position)
        result["custom_metrics"]["current_position_var"] = var
        result["custom_metrics"]["current_position_mean"] = mean
        del result["custom_metrics"]["current_position"]

    def on_learn_on_batch(
        self, *, policy: Policy, train_batch: SampleBatch, result: dict, **kwargs
    ) -> None:
        result["sum_actions_in_train_batch"] = train_batch["actions"].sum()


class CurriculumCapableSimpleCorridorEnv(SimpleCorridor, TaskSettableEnv):
    def __init__(self, config: EnvContext):
        SimpleCorridor.__init__(self, config)
        self.cur_level = 0

    @override(TaskSettableEnv)
    def get_task(self):
        """Implement this to get the current task (curriculum level)."""
        return self.cur_level

    @override(TaskSettableEnv)
    def set_task(self, task):
        """Implement this to set the task (curriculum level) for this env."""
        self.cur_level = task

def curriculum_fn(
    train_results: dict, task_settable_env: TaskSettableEnv, env_ctx: EnvContext
) -> TaskType:
    """Function returning a possibly new task to set `task_settable_env` to.
    """
    new_task = task_settable_env.cur_level
    return new_task


ray.init(num_cpus=14, num_gpus=1)
register_env(
    "curriculum_env", lambda config : CurriculumCapableSimpleCorridorEnv(config))

config = (
    PPOConfig()
    .environment(
        env="curriculum_env",
        env_config={"corridor_length": 5},
        env_task_fn=curriculum_fn,
    )
    .framework("torch")
    .training(
        model={"fcnet_hiddens": [32, 32]},
        train_batch_size=100_000,
    )
    .reporting(keep_per_episode_custom_metrics=True)
    .callbacks(callbacks_class=MyCallbacks)
    .rollouts(num_rollout_workers=12, num_envs_per_worker=2, enable_connectors=True)
    .resources(num_gpus=1, num_cpus_per_worker=1)
)

algo = config.build()

start = time.time()
for i in range(1, 21):
    algo.train()
    print(f"Iteration {i}, iter_time={time.time() - start:.2f}")
    start = time.time()
```

Here are the outputs I got:
- Without any changes:
```
ERROR:root:Curriculum update time 0.236
Iteration 1, iter_time=77.42
ERROR:root:Curriculum update time 0.440
Iteration 2, iter_time=79.76
ERROR:root:Curriculum update time 0.574
Iteration 3, iter_time=77.68
ERROR:root:Curriculum update time 0.678
Iteration 4, iter_time=81.45
ERROR:root:Curriculum update time 0.715
Iteration 5, iter_time=80.87
ERROR:root:Curriculum update time 0.755
Iteration 6, iter_time=79.87
ERROR:root:Curriculum update time 0.750
Iteration 7, iter_time=81.95
ERROR:root:Curriculum update time 0.784
Iteration 8, iter_time=81.68
ERROR:root:Curriculum update time 0.793
Iteration 9, iter_time=80.45
ERROR:root:Curriculum update time 0.803
Iteration 10, iter_time=80.18
ERROR:root:Curriculum update time 0.813
Iteration 11, iter_time=80.49
ERROR:root:Curriculum update time 0.817
Iteration 12, iter_time=81.26
ERROR:root:Curriculum update time 0.792
Iteration 13, iter_time=82.36
ERROR:root:Curriculum update time 0.868
Iteration 14, iter_time=82.09
ERROR:root:Curriculum update time 0.805
Iteration 15, iter_time=81.51
ERROR:root:Curriculum update time 0.815
Iteration 16, iter_time=81.29
ERROR:root:Curriculum update time 0.803
Iteration 17, iter_time=80.40
ERROR:root:Curriculum update time 0.814
Iteration 18, iter_time=80.12
ERROR:root:Curriculum update time 0.676
Iteration 19, iter_time=81.27
ERROR:root:Curriculum update time 0.736
Iteration 20, iter_time=81.25
```
- With suggestion `1` (using the shared-memory object-store):
```
ERROR:root:Curriculum update time 0.219
Iteration 1, iter_time=77.56
ERROR:root:Curriculum update time 0.172
Iteration 2, iter_time=77.67
ERROR:root:Curriculum update time 0.196
Iteration 3, iter_time=79.08
ERROR:root:Curriculum update time 0.217
Iteration 4, iter_time=81.12
ERROR:root:Curriculum update time 0.216
Iteration 5, iter_time=77.97
ERROR:root:Curriculum update time 0.237
Iteration 6, iter_time=78.88
ERROR:root:Curriculum update time 0.245
Iteration 7, iter_time=80.96
ERROR:root:Curriculum update time 0.247
Iteration 8, iter_time=79.95
ERROR:root:Curriculum update time 0.247
Iteration 9, iter_time=79.07
ERROR:root:Curriculum update time 0.250
Iteration 10, iter_time=80.31
ERROR:root:Curriculum update time 0.252
Iteration 11, iter_time=79.08
ERROR:root:Curriculum update time 0.246
Iteration 12, iter_time=81.98
ERROR:root:Curriculum update time 0.258
Iteration 13, iter_time=80.98
ERROR:root:Curriculum update time 0.252
Iteration 14, iter_time=82.42
ERROR:root:Curriculum update time 0.249
Iteration 15, iter_time=79.26
ERROR:root:Curriculum update time 0.257
Iteration 16, iter_time=80.78
ERROR:root:Curriculum update time 0.249
Iteration 17, iter_time=78.90
ERROR:root:Curriculum update time 0.254
Iteration 18, iter_time=78.97
ERROR:root:Curriculum update time 0.249
Iteration 19, iter_time=78.88
ERROR:root:Curriculum update time 0.097
Iteration 20, iter_time=75.56
```

- With suggestion `1 + 2` : 
```
ERROR:root:Curriculum update time 0.043
Iteration 1, iter_time=75.83
ERROR:root:Curriculum update time 0.091
Iteration 2, iter_time=76.58
ERROR:root:Curriculum update time 0.118
Iteration 3, iter_time=77.61
ERROR:root:Curriculum update time 0.136
Iteration 4, iter_time=77.74
ERROR:root:Curriculum update time 0.140
Iteration 5, iter_time=78.85
ERROR:root:Curriculum update time 0.147
Iteration 6, iter_time=78.42
ERROR:root:Curriculum update time 0.156
Iteration 7, iter_time=80.35
ERROR:root:Curriculum update time 0.164
Iteration 8, iter_time=79.53
ERROR:root:Curriculum update time 0.158
Iteration 9, iter_time=78.42
ERROR:root:Curriculum update time 0.166
Iteration 10, iter_time=80.44
ERROR:root:Curriculum update time 0.167
Iteration 11, iter_time=78.86
ERROR:root:Curriculum update time 0.162
Iteration 12, iter_time=78.96
ERROR:root:Curriculum update time 0.166
Iteration 13, iter_time=79.88
ERROR:root:Curriculum update time 0.168
Iteration 14, iter_time=79.03
ERROR:root:Curriculum update time 0.163
Iteration 15, iter_time=79.71
ERROR:root:Curriculum update time 0.156
Iteration 16, iter_time=79.17
ERROR:root:Curriculum update time 0.212
Iteration 17, iter_time=77.19
ERROR:root:Curriculum update time 0.257
Iteration 18, iter_time=78.37
ERROR:root:Curriculum update time 0.143
Iteration 19, iter_time=78.43
ERROR:root:Curriculum update time 0.161
Iteration 20, iter_time=78.96
```

Even on this simple use case, the slow-down was divided by around 5 (even though the slow-down was minor).

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

**Comment by ArturNiederfahrenhorst** [comment]
Marking this as P0 and selfassigning to check if this issue exists on RLlib v2 stack.

**Comment by ArturNiederfahrenhorst** [comment]
<img width="1700" height="1066" alt="Image" src="https://github.com/user-attachments/assets/2223ffbf-7a60-4139-b145-18cd3e8fa793" />

I translated the reproduction script to RLlib's v2 stack and the problem does not exist there.
Please take a look at the following if you are interested to see how the way we set this has changed:
https://github.com/ray-project/ray/blob/master/rllib/examples/curriculum/curriculum_learning.py#L190-L192

In your problem description, you make the point that the rfcs contain the entire results dict and that you were able to cut the latency of the operation by ray.putting the results first and recommend that we also only share a required subset of the results with the RolloutWorkers. The new APIs we use in the example above don't require any form of results dict anymore so this is not needed.

Thanks for raising the issue and feel free to re-open if there is more to discuss here.
